File: Hacker-Rank/Time-Complexity-Primality/primality.c
/*  Problem Statement: https://www.hackerrank.com/challenges/ctci-big-o/problem?h_l=interview&playlist_slugs%5B%5D=interview-preparation-kit&playlist_slugs%5B%5D=miscellaneous
    Author: striker
*/

#include<stdio.h>
#include<stdlib.h>
#include<inttypes.h>
#include<stdbool.h>
#include<assert.h>

static const bool check_primality(uint32_t);

int main(void) {
    uint8_t test;
    scanf("%"SCNu8, &test);
    assert(test > 0 && test < 31);
    while(test--) {
        uint32_t n;
        scanf("%"SCNu32, &n);
        assert(n > 0 && n < 2000000001);
        if(n == 1) {
            printf("Not prime\n");
        } else {
            printf("%s", check_primality(n) ? "Prime\n" : "Not prime\n");
        }
    }
    return EXIT_SUCCESS;
}

static const bool check_primality(uint32_t n) {
    bool is_prime = true;
    for(uint64_t i = 2; (i * i) <= n; ++i) {
        if(!(n % i)) {
            is_prime = false;
            break;
        }
    }
    return is_prime;
}

File: Hacker-Rank/Time-Complexity-Primality/primality.py
"""
	Problem Statement: https://www.hackerrank.com/challenges/ctci-big-o/problem
	Author: striker
"""

import math

def check_prime(number: int) -> bool:
	if number == 2 or number == 3:
		return True
	if number == 1 or not number % 2 or not number % 3:
		return False
	return not any(not number % d for d in range(2, 1 + int(math.sqrt(number))))

def main():
	for _ in range(int(input().strip())):
		print("Prime" if check_prime(int(input().strip())) else "Not prime")

if __name__ == "__main__":
	main()